src/pip/_internal/index.py

```diff
@@ -70,7 +102,62 @@ def _get_content_type(url, session):
     resp = session.head(url, allow_redirects=True)
     resp.raise_for_status()
 
-    return resp.headers.get("Content-Type", "")
+    _ensure_html_header(resp.headers)
+
+
+def _get_html_response(url, session):
+    """Access an HTML page with GET, and return the response.
+
+    This consists of three parts:
+
+    1. If URL scheme is file: and points to a directory, make it point to
+       index.html instead.
+    2. Actually perform the request. Raise HTTP exceptions on network failures.
+    3. Check whether Content-Type header to make sure the thing we got is HTML,
+       and raise `_NotHTML` if it's not.
+    """
+    logger.debug('Getting page %s', url)
+
+    # Tack index.html onto file:// URLs that point to directories
+    scheme, _, path, _, _, _ = urllib_parse.urlparse(url)
+    if (scheme == 'file' and os.path.isdir(urllib_request.url2pathname(path))):
```

Comment:
Can this check and URL conversion be moved to before the `try-except` and after the VCS check? That way you'd be passing to `_get_html_response()` the URL that you're actually requesting. It would also simplify the logic because this logic isn't needed to be wrapped in the `try-except`, so it would make everything easier to understand.

Reply:
Good eye! So are you suggesting moving it out of the try-except now, or postponing until a subsequent PR? Since it would both simplify the code a bit as well as fix an edge case (that in addition is probably extremely rare), it doesn't seem like there's any reason not to do it now.

Reply:
I think we can do it now, and maybe add a fragment to note the behavioural change for the off chance someone is actually depending on it.